src/services/signer.service.ts

```diff
@@ -222,4 +229,72 @@ export class SignerService {
 
     return await promise;
   }
+
+  /**
+   * If the ICRC-21 call to fetch the consent message fails, it might be due to the fact
+   * that the targeted canister does not implement the ICRC-21 specification.
+   *
+   * To address the potential lack of support for the most common types of calls for ledgers,
+   * namely transfer and approve, we use custom builders. Those builders construct
+   * messages similar to those that would be implemented by the canisters.
+   *
+   * @param {Object} params - The parameters for loading the consent message.
+   * @param {Omit<IcrcCallCanisterRequestParams, 'sender'>} params.params - The ICRC call canister parameters minus the sender.
+   * @param {SignerOptions} params.options - The signer options - host and owner.
+   * @returns {Promise<icrc21_consent_message_response>} - The consent message response.
+   * @throws The potential original error from the ICRC-21 call. The errors related to
+   *         the custom builder is ignored.
+   **/
+  private async loadConsentMessage(params: {
+    params: Omit<IcrcCallCanisterRequestParams, 'sender'>;
+    options: SignerOptions;
+  }): Promise<icrc21_consent_message_response> {
+    try {
+      return await this.callConsentMessage(params);
+    } catch (err: unknown) {
+      const fallbackMessage = await this.tryBuildConsentMessageOnError(params);
+
+      if ('Ok' in fallbackMessage) {
+        return fallbackMessage;
+      }
+
+      throw err;
+    }
+  }
+
+  private async tryBuildConsentMessageOnError({
+    params: {method, arg, canisterId},
+    options: {owner, host}
+  }: {
+    params: Omit<IcrcCallCanisterRequestParams, 'sender'>;
+    options: SignerOptions;
+  }): Promise<{NoFallback: null} | {Ok: icrc21_consent_info} | {Err: unknown}> {
+    const fn = SIGNER_BUILDERS[method];
+
+    if (isNullish(fn)) {
+      return {NoFallback: null};
+    }
+
+    try {
+      const tokenResponse = await this.#signerApi.ledgerMetadata({
+        params: {canisterId},
+        host,
+        owner
+      });
+
+      const token = mapTokenMetadata(tokenResponse);
+
+      if (isNullish(token)) {
+        return {Err: new Error('Incomplete token metadata.')};
```

Comment:
No need, I would say. This particular error also not handled in any specific way since it's the original error—the lack of ICRC-21 support—which will be bubbled up to the caller.

All error messages are currently provided in English. I'm not sure what criteria would be used for selecting another language. Maybe something to improve globally in the future, in that sense good point!